piq/fsim.py

```diff
@@ -0,0 +1,453 @@
+        scales: Number of wavelets used for computation of phase congruensy maps
+        orientations: Number of filter orientations used for computation of phase congruensy maps
+        min_length: Wavelength of smallest scale filter
+        mult: Scaling factor between successive filters
+        sigma_f: Ratio of the standard deviation of the Gaussian describing the log Gabor filter's
+            transfer function in the frequency domain to the filter center frequency.
+        delta_theta: Ratio of angular interval between filter orientations and the standard deviation
+            of the angular Gaussian function used to construct filters in the frequency plane.
+        k: No of standard deviations of the noise energy beyond the mean at which we set the noise
+            threshold  point, below which phase congruency values get penalized.
+
+    Shape:
+        - Input: Required to be 2D (H, W), 3D (C,H,W), 4D (N,C,H,W) or 5D (N,C,H,W,2), channels first.
+        - Target: Required to be 2D (H, W), 3D (C,H,W), 4D (N,C,H,W) or 5D (N,C,H,W,2), channels first.
+
+    Examples::
+
+        >>> loss = FSIMLoss()
+        >>> prediction = torch.rand(3, 3, 256, 256, requires_grad=True)
+        >>> target = torch.rand(3, 3, 256, 256)
+        >>> output = loss(prediction, target)
+        >>> output.backward()
+
+    References:
+        .. [1] Anish Mittal et al. "No-Reference Image Quality Assessment in the Spatial Domain",
+        https://live.ece.utexas.edu/publications/2012/TIP%20BRISQUE.pdf
+        """
+    def __init__(self, data_range: Union[int, float] = 1., reduction: str = 'mean', scales: int = 4,
+                 orientations: int = 4, min_length: int = 6, mult: int = 2, sigma_f: float = 0.55,
+                 delta_theta: float = 1.2, k: float = 2.0) -> None:
+
+        super().__init__()
+        self.data_range = data_range
+        self.reduction = reduction
+
+        # Save function with predefined parameters, rather than parameters themself
+        self.fsim = functools.partial(
+            fsim,
+            data_range=data_range,
+            reduction=reduction,
+            scales=scales,
+            orientations=orientations,
+            min_length=min_length,
+            mult=mult,
+            sigma_f=sigma_f,
+            delta_theta=delta_theta,
+            k=k,
+        )
+
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of FSIM as a loss function.
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+        Returns:
+            Value of FSIM loss to be minimized. 0 <= FSIM <= 1.
+        """
+        # All checks are done inside fsim function
+        score = self.fsim(prediction, target)
+
+        # Make sure value to be in [0, 1] range and convert to loss
+        return 1 - torch.clamp(score, 0, 1)
```

Comment:
Just curious whether we need to do something about the case when `torch.clamp(score, 0, 1)` is equal to 1 🤔 
Are there any cases then it may cause some problems?